sklearn/cluster/tests/test_optics.py

```diff
@@ -0,0 +1,386 @@
+    X = [[1, 1]]
+    clust = OPTICS(max_bound=5.0 * 0.3, min_samples=10)
+
+    # Run the fit
+    assert_raises(ValueError, clust.fit, X)
+
+
+def test_empty_extract():
+    # Test extract where fit() has not yet been run.
+    clust = OPTICS(max_bound=5.0 * 0.3, min_samples=10)
+    assert_raises(ValueError, clust.extract_dbscan, 0.01)
+
+
+def test_bad_extract():
+    # Test an extraction of eps too close to original eps
+    centers = [[1, 1], [-1, -1], [1, -1]]
+    X, labels_true = make_blobs(n_samples=750, centers=centers,
+                                cluster_std=0.4, random_state=0)
+
+    # Compute OPTICS
+    clust = OPTICS(max_bound=5.0 * 0.003, min_samples=10)
+    clust2 = clust.fit(X)
+    assert_raises(ValueError, clust2.extract_dbscan, 0.3)
+
+
+def test_close_extract():
+    # Test extract where extraction eps is close to scaled epsPrime
+
+    centers = [[1, 1], [-1, -1], [1, -1]]
+    X, labels_true = make_blobs(n_samples=750, centers=centers,
+                                cluster_std=0.4, random_state=0)
+
+    # Compute OPTICS
+    clust = OPTICS(max_bound=1.0, min_samples=10)
+    clust3 = clust.fit(X)
+    # check warning when centers are passed
+    assert_warns(RuntimeWarning, clust3.extract_dbscan, .3)
+    # Cluster ordering starts at 0; max cluster label = 2 is 3 clusters
+    assert_equal(max(clust3.extract_dbscan(.3)[1]), 2)
+
+
+def test_auto_extract_hier():
+    # Tests auto extraction gets correct # of clusters with varying density
+
+    # Generate sample data
+    rng = np.random.RandomState(0)
+    n_points_per_cluster = 250
+
+    C1 = [-5, -2] + .8 * rng.randn(n_points_per_cluster, 2)
+    C2 = [4, -1] + .1 * rng.randn(n_points_per_cluster, 2)
+    C3 = [1, -2] + .2 * rng.randn(n_points_per_cluster, 2)
+    C4 = [-2, 3] + .3 * rng.randn(n_points_per_cluster, 2)
+    C5 = [3, -2] + 1.6 * rng.randn(n_points_per_cluster, 2)
+    C6 = [5, 6] + 2 * rng.randn(n_points_per_cluster, 2)
+    X = np.vstack((C1, C2, C3, C4, C5, C6))
+
+    # Compute OPTICS
+
+    clust = OPTICS(min_samples=9)
+
+    # Run the fit
+    clust.fit(X)
+
+    assert_equal(len(set(clust.labels_)), 6)
+
+
+def test_cluster_pruning():
+    # Tests pruning left and right, deletion of empty node_lists
+    # Parameters chosen specifically for this task
+
+    # Three pseudo clusters, 2 of which are too small
+    reach = np.array([np.inf, 0.9, 0.9, 1.0, 0.89, 0.88, 10, .9, .9, .9,
+                      10, 0.9, 0.9, 0.89, 0.88, 10, .9, .9, .9, .9])
+    # Normalize
+    reach = reach / np.max(reach[1:])
+
+    ordering = np.r_[0:20]
+    cluster_boundaries = _find_local_maxima(reach, 5)
+    root = _TreeNode(ordering, 0, 20, None)
+
+    # Build cluster tree inplace on root node
+    _cluster_tree(root, None, _find_local_maxima(reach, 5),
+                  reach, ordering, 5, .75, .7, .4, .3)
+    assert_equal(root.split_point, cluster_boundaries[0])
+    assert_equal(2, len(root.children))
+
+
+def test_sigmin_pruning():
```

Comment:
It is hard to see the differences between this and the previous test. Could they share more code, or be merged?

Reply:
I'm open to suggestions on how to condense here. Since the unit tests are setup as functions with no arguments, specifying a different data input seems to require a mostly duplicate test. This test and the one prior test for different points of coverage in auto_extract that are activated with different inputs.

Reply:
pytest.mark.parametrize takes care of avoiding duplicate tests, to some extent at least.

Reply:
added this and condensed these two tests to one.